x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/TrainedModelStatsService.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.inference;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.support.PlainActionFuture;
+import org.elasticsearch.action.support.WriteRequest;
+import org.elasticsearch.action.update.UpdateRequest;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.component.LifecycleListener;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentFactory;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.script.ScriptType;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xpack.core.ml.MlStatsIndex;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceStats;
+import org.elasticsearch.xpack.core.ml.utils.ToXContentParams;
+import org.elasticsearch.xpack.ml.MachineLearning;
+import org.elasticsearch.xpack.ml.utils.persistence.ResultsPersisterService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+
+public class TrainedModelStatsService {
+
+    private static final Logger logger = LogManager.getLogger(TrainedModelStatsService.class);
+
+    // Script to only update if stats have increased since last persistence
+    private static final String STATS_UPDATE_SCRIPT = "" +
+        "if (ctx._source.missing_all_fields_count >= params.missing_all_fields_count\n" +
```

Comment:
I don't see how an update could be queued with at least one field having changed am I missing something?

Reply:
This is to protect against non-monotonically increasing stats. 

There is a pathological scenario where:
 - model is loaded
 - writes stats
 - unloaded
 - reloaded before previous stats write occurs and starts at the previous values
 - attempts to write stats again, but the new values could be lower than the previously written ones. 

Scenarios like this one are where I wish I could use optimistic concurrency, but since the underlying indices are ILM managed, I don't know when an index switches out from under the process :/. This FEELS like a fundamental flaw with optimistic concurrency and ILM. I really wish there was a `POST` where it either HAS to match the seq_no and primary_term OR if the doc doesn't exist, create it.

Reply:
This complexity makes me think that one of the following might have to be done

- stats should be append only for inference some how (`sum` aggs to pull them together). This feels like we could have tons of docs being written to the .ml-stats index, but is the easiest solution for now...

- we don't use the .ml-stats index and do something similar to transforms (keeping track of the original underlying index name). This adds complexity as we now have yet another index, alias woes, and we will have to add optimistic concurrency controls between the stats persister and the thing persisting the stats.

Reply:
Silly question why can't you increment the fields in the upsert, that way it doesn't matter if the updates arrive out of order. The stats accumulator would have to change track only the changes since the last persist but that is simple enough

```ctx._source.missing_all_fields_count += params.missing_all_fields_count;```

Reply:
🤔

I will work on it. I think this SHOULD work, my original design was not assuming a single input queue. Here, we do have a single persistence queue that guarantees order.

Reply:
The counters are not thread safe when reset to `0`

The way to address this would be to keep track of the inference stats in `LocalModel` with an atomic reference. Then when stats are persisted to swap it out with a new one. I am not 100% of the performance impact. 

Right now performance is not really measured. So 🤷‍♂